File: solutions/javascript/robot-name/1/robot-name.js
const LETTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const DIGITS = '0123456789';

export class Robot {
    /** @type {string[]} */
    static usedNames = [];

    constructor() {
        this.reset();
    }

    get name() {
        return this._name;
    }

    reset() {
        do {
            this._name = this.generateName();
        } while (Robot.usedNames.includes(this._name));
        Robot.usedNames.push(this._name);
    }

    generateName() {
        let name = '';
        for (let i = 0; i < 2; i++) {
            name += LETTERS.charAt(Math.floor(Math.random() * LETTERS.length));
        }
        for (let i = 0; i < 3; i++) {
            name += DIGITS.charAt(Math.floor(Math.random() * DIGITS.length));
        }
        return name;
    }
}

Robot.releaseNames = () => {};

File: solutions/go/robot-name/1/robot_name.go
package robotname

import "math/rand"

type Robot struct {
	name string
}

var usedNames = map[string]bool{}

func (r *Robot) Name() (string, error) {
	if r.name != "" {
		return r.name, nil
	}
	r.name = generateRandomName()
	for usedNames[r.name] {
		r.name = generateRandomName()
	}
	usedNames[r.name] = true
	return r.name, nil
}

func (r *Robot) Reset() {
	if r.name != "" {
		delete(usedNames, r.name)
		r.name = ""
	}
}

func generateRandomName() string {
	name := []rune{}
	for i := 0; i < 2; i++ {
		name = append(name, 'A'+rune(rand.Intn(26)))
	}
	for i := 0; i < 3; i++ {
		name = append(name, '0'+rune(rand.Intn(10)))
	}
	return string(name)
}
